crates/bootstrap_srv/src/store.rs

```diff
@@ -0,0 +1,191 @@
+    pub fn parse(&self) -> std::io::Result<crate::ParsedEntry> {
+        let mut tmp = Vec::with_capacity(self.length);
+        self.read(&mut tmp)?;
+        crate::ParsedEntry::from_slice(&tmp)
+    }
+}
+
+/// Tempfile-based virtual memory solution.
+pub struct Store {
+    writer_pool: WriterPool,
+}
+
+impl Default for Store {
+    fn default() -> Self {
+        Self {
+            writer_pool: Arc::new(
+                Mutex::new(std::collections::VecDeque::new()),
+            ),
+        }
+    }
+}
+
+impl Store {
+    /// Write an entry to the virtual memory store, getting back
+    /// a reference that will allow future reading.
+    pub fn write(&self, content: &[u8]) -> std::io::Result<StoreEntryRef> {
+        use std::io::{Seek, Write};
+        let mut writer = {
+            match self.writer_pool.lock().unwrap().pop_front() {
+                Some(writer) => writer,
+                None => {
+                    let read_clone = Arc::new(tempfile::NamedTempFile::new()?);
+                    let writer = read_clone.reopen()?;
+                    Writer { writer, read_clone }
+                }
+            }
+        };
+        let read_clone = writer.read_clone.clone();
+        let offset = writer.writer.stream_position()?;
+        let length = content.len();
+
+        writer.writer.write_all(content)?;
+        writer.writer.sync_data()?;
+
+        if offset + (length as u64) < MAX_PER_TEMPFILE {
+            self.writer_pool.lock().unwrap().push_back(writer);
+        }
+
+        Ok(StoreEntryRef {
+            read_clone,
+            offset,
+            length,
+        })
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn happy_sanity() {
+        let s = Store::default();
+
+        let hello = s.write(b"Hello ").unwrap();
+        let world = s.write(b"world!").unwrap();
+
+        let mut buf = Vec::new();
+        hello.read(&mut buf).unwrap();
+        world.read(&mut buf).unwrap();
+
+        assert_eq!(b"Hello world!", buf.as_slice());
+    }
+
+    #[test]
+    fn happy_multi_thread_sanity() {
+        const COUNT: usize = 10;
+        let mut all = Vec::with_capacity(COUNT);
+
+        let s = Arc::new(Store::default());
+        let b = Arc::new(std::sync::Barrier::new(COUNT));
+
+        let (send, recv) = std::sync::mpsc::channel();
+
+        for i in 0..COUNT {
+            let send = send.clone();
+            let s = s.clone();
+            let b = b.clone();
+
+            all.push(std::thread::spawn(move || {
+                b.wait();
+                let wrote = format!("index:{i}");
+                let r = s.write(wrote.as_bytes()).unwrap();
+                send.send((wrote, r)).unwrap();
+            }));
+        }
+
+        for _ in 0..COUNT {
+            let (wrote, r) = recv.recv().unwrap();
+            let mut read = Vec::new();
+            r.read(&mut read).unwrap();
+            assert_eq!(wrote.as_bytes(), read);
+        }
+    }
+}
```

Comment:
I assume we want a lot testing more here, but i assume you know that :)

Reply:
@mattyg - I'm considering the testing in [Testing #11](https://github.com/holochain/kitsune2/pull/11) as sufficient for covering this. If you don't agree, please let me know what else you'd like to see here before approving : )

Reply:
whoops my bad